### ISSUE TITLE
Area and room restriction per user

### ISSUE THREAD

**Issue by mauroman77** [meeting-room-booking-system/mrbs-code#3585]
**Is your feature request related to a problem? Please describe.**
Hello, I don't have a problem, I would like to know if the following functionality can be implemented.
It would help me to restrict user access to certain areas and even with access to some area, restrict access to certain rooms.

**Describe the solution you'd like**
I imagine a table that stores user_id, area_id and room_id where this data is stored and then that user only has access to create events to the pre-established areas and the indicated rooms.

**Describe alternatives you've considered**
I imagine on the edit_users.php screen below the email input a fieldset per area with a checkbox to enable the area and a checkbox for each room that exists in that area

**Additional context**
The idea is that the users will be professors at a university who must reserve an exam date, for this I have decided to use the areas as academic careers and the rooms as the year of study (e.g. first year, second year... etc.)
I will be very grateful in advance for any help you can give me.

**Comment by NolanCrochot** [comment]
I was just about to create the exact same thread. I tried to edit it by myself, but as a front-end developer with few back-end knowledge, I'm a bit lost.
It would be nice if we can assign members to (a) group(s), and show rooms to specific groups only.

**Comment by campbell-m** [comment]
This is the purpose of [the roles branch](https://github.com/meeting-room-booking-system/mrbs-code/tree/roles).  It introduces the concepts of groups and roles and allows you to assign roles to users and groups.  Each role allows you to define permissions (read, write and admin) for rooms and areas.

You are welcome to try it out, but be warned that (a) it is still under development, (b) it's only been tested so far with the 'db' and 'ldap' authentication types and (c) it will upgrade your database in a non-backwards compatible manner.  So you'd probably want to test it in isolation until you are comfortable with it.

**Comment by mauroman77** [comment]
> This is the purpose of [the roles branch](https://github.com/meeting-room-booking-system/mrbs-code/tree/roles). It introduces the concepts of groups and roles and allows you to assign roles to users and groups. Each role allows you to define permissions (read, write and admin) for rooms and areas.
> 
> You are welcome to try it out, but be warned that (a) it is still under development, (b) it's only been tested so far with the 'db' and 'ldap' authentication types and (c) it will upgrade your database in a non-backwards compatible manner. So you'd probably want to test it in isolation until you are comfortable with it.

What role must the user have for the role configuration to take effect?

**Comment by mauroman77** [comment]
What am I missing to configure so that the group checkboxes are enabled?
Attached image
![Sistema-de-reserva-de-salas](https://github.com/meeting-room-booking-system/mrbs-code/assets/5404643/8be89ff9-8106-437b-a570-a681493e9384)

**Comment by campbell-m** [comment]
> What role must the user have for the role configuration to take effect?

The user must be an MRBS admin in order to be able to configure roles.  The term "admin" in the room/area permissions above is potentially confusing.  If you have admin (actually it's called "All" at the moment) rights for a room/area it means that you are able to edit and delete other people's bookings in that room/area.

**Comment by campbell-m** [comment]
> What am I missing to configure so that the group checkboxes are enabled?

Sorry, I think that bit hasn't been implemented yet for the 'db' auth type.  It will recognise LDAP groups if you are using LDAP.  But  for the moment for the 'db' auth type you can just assign roles to users.

As I say, it's still under development and I need to get back to it ...

**Comment by mauroman77** [comment]
Hello, sorry for the inconvenience, the room restriction capabilities are not implemented yet???, I have decided to ignore the use of groups and focus on the room restriction, but I cannot get the role functionality to work, if it does not work yet It is finished, could you tell me how you recommend continuing with your work, or better yet, if you plan to finish, when would it be?
Thank you very much and again sorry for the inconvenience.

**Comment by campbell-m** [comment]
Yes, it should work.  What have you done and what is not working?

**Comment by mauroman77** [comment]
Hello, in the attached file I indicate step by step what I am doing.
[test role.pdf](https://github.com/meeting-room-booking-system/mrbs-code/files/13719646/test.role.pdf)

**Comment by campbell-m** [comment]
You need to set the user's privileges to "user", not "none".

**Comment by mauroman77** [comment]
Hello, thanks for the information, I have tested the role configuration and it works very well, I want to bother you with one more query, there are context variables, if I want to know the id of the user who is interacting with the edit_entry.php file, also It would help me to be able to access the context variable to read the area id and room id, thank you very much in advance

**Comment by mauroman77** [comment]
Hello, I noticed that the area and room data are in the url and to be able to use them in the edit_entry.php file I use the following lines
$area_id = isset($_GET['area']) ? intval($_GET['area']) : 0;
$id_room = isset($_GET['room']) ? intval($_GET['room']) : 0;
I would be missing knowing the ID of the user who is consulting the page

**Comment by campbell-m** [comment]
The area and room ids are always available as the global variables `$area` and `$room`.  You can get the username as follows:

~~~php
$mrbs_user = session()->getCurrentUser();
$mrbs_username = (isset($mrbs_user)) ? $mrbs_user->username : null;
~~~

But what are you trying to do anyway?  If you are trying to restrict some users' access to particular rooms then there are better and easier ways of doing so.

**Comment by sauadmin** [comment]
Hello! For bookings yet to be approved, how can we change the ? char on the calendar display to something else such as a prefix of [TO BE APPROVED] or something of that sort to make it more clear that the booking has'nt been approved as yet.

**Comment by campbell-m** [comment]
> Hello! For bookings yet to be approved, how can we change the ? char on the calendar display to something else such as a prefix of [TO BE APPROVED] or something of that sort to make it more clear that the booking has'nt been approved as yet.

In future please could you raise a new issue for a new question.  I have done this for you at Issue #3599 and answered it there.

**Comment by campbell-m** [comment]
> Is it possible to also have view rights along with read/write/alll. I would like to be able to restrict users to not even view some rooms so they don't have access to information that other users have included in their bookings?

There are two things you could do:

1. Deny certain usrs read rights to certain rooms.  Then they won't even be able to view the rooms.
2. Make bookings private.  This means that only the owner of the booking and admins can see certain details of the booking.  See the section on private bookings in systemdefaults.inc.php.

**Comment by antham1616** [comment]
hi, how do you actually setup/start using the roles? and what are the requirements

**Comment by campbell-m** [comment]
> hi, how do you actually setup/start using the roles? and what are the requirements

Have you installed the roles branch?  If so you should see two new menu entries, "Roles" and "Groups".
